### PR TITLE
add osx build

### PR DESCRIPTION
**Pull request by ltalirz**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

Not saying that I'll make it work - just starting to look what dependencies would be missing & what would need to be fixed (it seems branches on forks aren't built on CI, one needs to make a PR for this).

### CODE REVIEW

**Comment by ltalirz** [comment]
@conda-forge-admin , please rerender

**Comment by ltalirz** [comment]
@oschuett @jan-janssen 
One question - should one try to use the default C/C++ compiler and linker on MacOS (=clang) or should one force to use the gnu one?

I'm now running into issues with missing symbols at the linking stage https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=241443&view=logs&j=74807954-f6b3-568c-328d-43d46f211bbd&t=29da2991-df68-5dd0-9e2b-93a312f5b49e&l=1835
```
$SRC_DIR/tools/build_utils/fypp -n $SRC_DIR/src/input/input_keyword_types.F input_keyword_types.F90
ld: warning: could not create compact unwind for _MAIN__: stack subq instruction is too different from dwarf stack size
$BUILD_PREFIX/bin/x86_64-apple-darwin13.4.0-gfortran -c -march=core2 -mtune=haswell -ftree-vectorize -fPIC -fstack-protector -O2 -pipe -isystem $PREFIX/include -fdebug-prefix-map=$SRC_DIR=/usr/local/src/conda/cp2k-7.1.0 -fdebug-prefix-map=$PREFIX=/usr/local/src/conda-prefix -ffree-form -D__COMPILE_ARCH="\"Darwin-x86-64-conda\"" -D__COMPILE_DATE="\"Tue Nov 17 16:26:14 UTC 2020\"" -D__COMPILE_HOST="\"Mac-1605629663877.local\"" -D__COMPILE_REVISION="\"git:e635599\"" -D__DATA_DIR="\"$SRC_DIR/data\"" -D__SHORT_FILE__="\"input/input_keyword_types.F\"" -I'$SRC_DIR/src/input/' -I'$SRC_DIR/obj/Darwin-x86-64-conda/sopt/exts/dbcsr' input_keyword_types.F90 
Undefined symbols for architecture x86_64:
```

**Comment by jan-janssen** [comment]
@ltalirz As all the dependencies are compiled with the default compiler (=clang) I believe that is the more promising direction.

**Comment by ltalirz** [comment]
> @ltalirz As all the dependencies are compiled with the default compiler (=clang) I believe that is the more promising direction.

Does this apply to the fortran compiler as well?
I notice the `gfortran` is the result of `{{ compiler('fortran') }}`  and is thus probably somewhat recommended by conda-forge (?); at the same time one could also imagine using `flang`(?).
Has there been any testing of cp2k with the llvm family?

And when using the clang `ld` it turns out I need to link in libgfortran and libc explicitly.
Just wondering whether this is the right path; it would be nice if someone could give a look to the [ARCH file](https://github.com/conda-forge/cp2k-feedstock/pull/7/files#diff-bad75dc3d01945420eb0d2100da0330fce7c8852f447e298a517dacff7f2c83f) and let me know whether this makes sense at all.

**Comment by jan-janssen** [comment]
Yes, `gfortran` is the default Fortran compiler for Mac Os X on conda https://docs.conda.io/projects/conda-build/en/latest/resources/compiler-tools.html#compiler-packages . This means all libraries are compiled with `gfortran` and linking with a different compiler can cause issues. I have to admit I never tested `flang` in detail, I just remember that `ifort` is a lot of trouble when mixing it with `gfortran`. That is all I can say from the conda side, @oschuett is the cp2k expert, maybe he can take a look at the `ARCH` file.

**Comment by oschuett** [comment]
I've no experience with Mac OS, but `flang` can [not build](https://www.cp2k.org/dev:compiler_support) CP2K yet.

**Comment by ltalirz** [comment]
@jan-janssen @oschuett  build and `cp2k.sopt -h` pass, but the archfile should still be cleaned up.
Please give it a review, and check whether it looks reasonable in general.

There's still a couple of [linker warnings](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=246355&view=logs&j=d5375eb1-1d95-5809-042d-5185ca8a7328&t=5dd76214-332f-5e8e-dfa5-5354af1e90dc&l=3579) of the form
```
ld: warning: could not create compact unwind for ___cp_dbcsr_cp2k_link_MOD_create_dbcsr_section: stack subq instruction is too different from dwarf stack size
```
Not sure how important they are, [might](https://github.com/DNS-OARC/dnscap/issues/39) have to do with the version of libbind ?

Plus [a couple of warnings about unused packages](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=246355&view=logs&j=d5375eb1-1d95-5809-042d-5185ca8a7328&t=5dd76214-332f-5e8e-dfa5-5354af1e90dc&l=4957):
```
WARNING (cp2k): run-exports library package conda-forge::liblapack-3.8.0-11_h43114d0_netlib in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
WARNING (cp2k): run-exports library package conda-forge::libblas-3.8.0-11_h43114d0_netlib in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`
WARNING (cp2k): dso library package conda-forge::libcxx-11.0.0-h4c3b8ed_1 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
WARNING (cp2k): run-exports library package conda-forge::libgfortran-4.0.0-h50e675f_13 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
```

File: recipe/build.sh
Comment (ltalirz) [inline recipe/build.sh]: wasn't able to find a generic conda variable that just tells the OS platform - if you happen to know one, let me know.
Otherwise one could also check whether this particular variable is non-empty

File: recipe/meta.yaml
Comment (ltalirz) [inline recipe/meta.yaml:17]: (or ` # [ win ]`)

**Comment by ltalirz** [comment]
After re-enabling `make test` it [fails with](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=246372&view=logs&j=74807954-f6b3-568c-328d-43d46f211bbd&t=29da2991-df68-5dd0-9e2b-93a312f5b49e&l=4756):
```
/Users/runner/miniforge3/conda-bld/cp2k_1606777594219/work/tools/regtesting/do_regtest: line 168: syntax error near unexpected token `&'
```
I guess it's this line https://github.com/cp2k/cp2k/blob/971f4a76691299ddc5202670f1b2667c7623409c/tools/regtesting/do_regtest#L166

Maybe that's a case of a wrong shell being used?
Hm...

**Comment by ltalirz** [comment]
Looking back at the corresponding line in the regtest script, the `;&` almost looks like a typo to me (should be `;;`?)
https://github.com/cp2k/cp2k/blob/971f4a76691299ddc5202670f1b2667c7623409c/tools/regtesting/do_regtest#L166

@dev-zero if I see correctly, git seems to blames you for adding these lines, could you enlighten me?

I'm also wondering why I see this triggered on the macos build only; perhaps some software is missing that results in the script ending up at this point, while it never reaches there on the linux build?

**Comment by dev-zero** [comment]
> @dev-zero if I see correctly, git seems to blames you for adding these lines, could you enlighten me?

That's a fallthrough statement introduced in bash 4. If it fails there you are either using a very old bash (bash 4 was released in 2009) or a not fully compatible bash substitute.
I used it as a replacement to maintain the same options but give users who were still using the `-svn` option a notice that they should switch over. I am going to remove it now.

**Comment by jan-janssen** [comment]
@ltalirz I included the changes from @dev-zero in your pull request as a patch - once the tests pass we can merge this pull request.

**Comment by jan-janssen** [comment]
```
2020-12-02T13:27:13.9787160Z /Users/runner/miniforge3/conda-bld/cp2k_1606914656707/work/tools/regtesting/do_regtest: line 202: nproc: command not found
2020-12-02T13:27:13.9937290Z ps: %r: keyword not found
2020-12-02T13:27:13.9937960Z ps: no valid keywords; valid keywords:
2020-12-02T13:27:13.9993650Z ERROR: maxtasks should be greater or equal numprocs*OMP_NUM_THREADS
2020-12-02T13:27:13.9994180Z        maxtasks        = 
2020-12-02T13:27:13.9994520Z        numprocs        = 1
2020-12-02T13:27:13.9994840Z        OMP_NUM_THREADS = 1
2020-12-02T13:27:14.0074940Z make[2]: *** [/Users/runner/miniforge3/conda-bld/cp2k_1606914656707/work/Makefile:138: test] Error 8
2020-12-02T13:27:14.0078920Z make[1]: *** [/Users/runner/miniforge3/conda-bld/cp2k_1606914656707/work/Makefile:123: sopt] Error 2
2020-12-02T13:27:14.0082340Z make: *** [Makefile:118: test] Error 2
2020-12-02T13:27:15.2915680Z Traceback (most recent call last):
```

`nproc` is not available on mac os X, it should be `sysctl -n hw.logicalcpu` as far as I know.

**Comment by dev-zero** [comment]
wrt the compiler selection: I believe we have some people who build the C/C++ parts with clang and the rest with gfortran. On the other hand since gfortran calls gcc's C preprocessor I would rather recommend to also build the C/C++ parts with gcc if possible for consistency.
Please add `-fbacktrace`: since we're using `abort()` on user input errors this may be crucial to figure out the codepath which triggered it. And also enable `-fopenmp` on all languages. Starting from v8.1 we will not supported single-thread versions anymore.

**Comment by jan-janssen** [comment]
@conda-forge-admin, please rerender

**Comment by ltalirz** [comment]
Thanks a lot for the fixes @jan-janssen , looks like we're getting close!

@dev-zero Would you be able to give a quick review of the ARCH file; pointing out any issues you see?

Apart from this, there is only the rather pedestrian way how I'm currently [triggering the osx build](https://github.com/conda-forge/cp2k-feedstock/pull/7/files#diff-d7075654874cb08007a21aaab3ecd4b3453a9087e7505d034d548b8938b599bcR4) - I can replace this with a check whether the `MACOSX_DEPLOYMENT_TARGET` variable is set so that at least it doesn't fail as soon as there is an osx update.
If someone has a better suggestion, please let me know.

**Comment by ltalirz** [comment]
> wrt the compiler selection: I believe we have some people who build the C/C++ parts with clang and the rest with gfortran. On the other hand since gfortran calls gcc's C preprocessor I would rather recommend to also build the C/C++ parts with gcc if possible for consistency.

Shall we give clang a try, and switch to gcc if it fails?

At least the [regtest results](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=247105&view=logs&j=58ac6aab-c4bc-5de2-2894-98e408cc8ec9&t=933f325c-924e-533d-4d95-e93b5843ce8b&l=8484) seem to look fine.

> And also enable `-fopenmp` on all languages. Starting from v8.1 we will not supported single-thread versions anymore.

If it's ok, I will add this in a second PR, since this also affects the linux build.

**Review by jan-janssen — APPROVED** [review]
Looks good to me. @ltalirz do you want to add yourself as a maintainer? @oschuett and @dev-zero any further comments? Otherwise I merge this pull request and we can focus on the `openmp` part in a different pull request.

**Comment by ltalirz** [comment]
Ok for me to proceed; I can add myself also in the next PR

**Comment by ltalirz** [comment]
By the way, what's the policy for adding the openmp version - since the filename for the openmp executable will change from .sopt to .ssmp (to be consistent with cp2k's general approach), removing the .sopt executable would be a "breaking change".
I.e. do we simply build *both* sopt and ssmp version for the time being?

**Comment by jan-janssen** [comment]
> By the way, what's the policy for adding the openmp version - since the filename for the openmp executable will change from .sopt to .ssmp (to be consistent with cp2k's general approach), removing the .sopt executable would be a "breaking change".
> I.e. do we simply build _both_ sopt and ssmp version for the time being?

I primarily build this package for https://github.com/conda-forge/pycp2k-feedstock so I can update the link there, still it is possible that others are using it as well, so maybe it makes sense to only do the switch once the new version is released. But I am fine either way.

**Comment by oschuett** [comment]
LGTM

Thanks Leo! It's good to have regular OSX builds.

**Pull request merged**
